overlord/fdestate/extop.go

```diff
@@ -0,0 +1,206 @@
+	case PreparingStatus:
+		return "preparing"
+	case DoingStatus:
+		return "doing"
+	case DoneStatus:
+		return "done"
+	case CompletingStatus:
+		return "completing"
+	case AbortingStatus:
+		return "aborting"
+	case ErrorStatus:
+		return "error"
+	}
+	panic(fmt.Sprintf("internal error: unexpected external operation status code: %d", s))
+}
+
+// externalOperation captures an externally running operation that is tracked in
+// the state.
+type externalOperation struct {
+	Kind string `json:"kind"`
+	// ChangeID is ID of a state.Change associated with the operation.
+	ChangeID string                  `json:"change-id"`
+	Status   ExternalOperationStatus `json:"status"`
+	// Err when not empty, carries the error message, usually associated with
+	// the operation having an ErrorStatus or AbortingStatus.
+	Err string `json:"err"`
+	// Context is an opaque piece of data associated with the operation.
+	Context json.RawMessage `json:"context"`
+}
+
+// SetFailed is a convenience helper, which essentially does
+// SetStatus(ErrorStatus), but also sets the error description.
+func (e *externalOperation) SetFailed(msg string) {
+	e.Status = ErrorStatus
+	e.Err = msg
+}
+
+func (e *externalOperation) SetStatus(status ExternalOperationStatus) {
+	e.Status = status
+}
+
+// Equal checks the operations for equality by comparing their kind and the
+// parent change ID.
+func (e *externalOperation) Equal(other *externalOperation) bool {
+	return e.ChangeID == other.ChangeID && e.Kind == other.Kind
+}
+
+// IsReady returns true when operation status is one of the final ones
+// (ErrorStatus or DoneStatus).
+func (e *externalOperation) IsReady() bool {
+	switch e.Status {
+	case DoneStatus, ErrorStatus:
+		return true
+	}
+
+	return false
+}
+
+// findFirstPendingExternalOperationByKind attempts to find a first instance of
+// an operation in the state with a given kind. If none is found, nil is
+// returned, without any errors.
+func findFirstPendingExternalOperationByKind(st *state.State, kind string) (foundOp *externalOperation, err error) {
+	err = withFdeState(st, func(fde *FdeState) (modified bool, err error) {
+		for _, op := range fde.PendingExternalOperations {
+			if op.Kind == kind && !op.IsReady() {
+				foundOp = &op
+				break
+			}
+		}
+		return false, nil
+	})
+	return foundOp, err
+}
+
+// findExternalOperationByChangeID looks up an operation tracked in the state
+// with a given change ID
+func findExternalOperationByChangeID(st *state.State, changeID string) (foundOp *externalOperation, err error) {
+	err = withFdeState(st, func(fde *FdeState) (modified bool, err error) {
+		for _, op := range fde.PendingExternalOperations {
+			if op.ChangeID == changeID {
+				foundOp = &op
+				break
+			}
+		}
+		return false, nil
+	})
+	return foundOp, err
+}
+
+// addExternalOperation adds external operation to the state. Does not perform
+// any checks for duplicates
+func addExternalOperation(st *state.State, op *externalOperation) error {
```

Comment:
This helper doesn't have the same spirit of the other helpers of assuming only one op matching changeID/kind.

OTOH, If it's use is limited the doc comment should be enough.